conans/model/ref.py

```diff
@@ -67,20 +82,49 @@ def loads(text):
         text = ConanFileReference.whitespace_pattern.sub("", text)
         tokens = ConanFileReference.sep_pattern.split(text)
         try:
-            name, version, user, channel = tokens
+            if len(tokens) not in (4, 5):
+                raise ValueError
+            name, version, user, channel = tokens[0:4]
+            revision = tokens[4] if len(tokens) == 5 else None
         except ValueError:
             raise ConanException("Wrong package recipe reference %s\nWrite something like "
                                  "OpenCV/1.0.6@user/stable" % text)
-        return ConanFileReference(name, version, user, channel)
+        obj = ConanFileReference(name, version, user, channel)
+        obj.revision = revision
+        return obj
 
     def __repr__(self):
         return "%s/%s@%s/%s" % (self.name, self.version, self.user, self.channel)
 
+    def full_repr(self):
+        str_rev = "#%s" % self.revision if self.revision else ""
+        return "%s%s" % (str(self), str_rev)
+
+    def copy_with_revision(self, revision):
+        tmp = ConanFileReference.loads(str(self))
+        tmp.revision = revision
+        return tmp
+
+    def copy_without_revision(self):
+        ret = ConanFileReference.loads(str(self))
+        ret.revision = None
+        return ret
+
 
 class PackageReference(namedtuple("PackageReference", "conan package_id")):
     """ Full package reference, e.g.:
     opencv/2.4.10@lasote/testing, fe566a677f77734ae
     """
+    revision = None
+
+    def __new__(cls, conan, package_id, revision=None):
+        if revision:
+            ConanName.validate_name(revision)
```

Comment:
Why the revision is validated here and not in ConanFileReference? (could be ommitted here?)

Reply:
I don't know... Adding the validation to the ConanFileReference.